File: engine/src/main/java/org/epdb/engine/volcano/Selection.java
package org.epdb.engine.volcano;

import org.epdb.engine.comparison.Predicate;
import org.epdb.engine.dto.Tuple;

public record Selection(
        Predicate predicate,
        Operator childOperator
) implements Operator {

    @Override
    public void open() {
        this.childOperator.open();
        System.out.println("Opened child operator...");
    }

    @Override
    public Tuple next() {
        while (true) {
            var tuple = childOperator.next();

            if (tuple == null) {
                return null;
            }

            if (predicate.evaluate(tuple)) {
                return tuple;
            }
        }
    }

    @Override
    public void close() {
        this.childOperator.close();
        System.out.println("Closed child operator...");
    }

}

File: modules/engine/src/main/kotlin/org/epdb/engine/databaseoperator/Selection.kt
package org.epdb.engine.databaseoperator

import org.epdb.engine.comparison.Predicate
import org.epdb.engine.dto.Tuple
import org.epdb.org.epdb.commons.Logger

data class Selection(
    private val predicate: Predicate,
    private val childOperator: Operator
) : Operator {

    override fun open() {
        this.childOperator.open()
        Logger.info("Opened child operator...")
    }

    override fun next(): Tuple? {
        while (true) {
            val tuple = childOperator.next() ?: return null

            if (predicate.evaluate(tuple)) {
                return tuple
            }
        }
    }

    override fun close() {
        this.childOperator.close()
        Logger.info("Closed child operator...")
    }
}
